pkg/apis/pipeline/v1alpha1/param_types.go

```diff
@@ -0,0 +1,256 @@
+// Used in JSON unmarshalling so that a single JSON field can accept
+// either an individual string or an array of strings.
+type ParamValue struct {
+	Type      ParamType // Represents the stored type of ParamValues.
+	StringVal string
+	// +listType=atomic
+	ArrayVal  []string
+	ObjectVal map[string]string
+}
+
+// ArrayOrString is deprecated, this is to keep backward compatibility
+//
+// Deprecated: Use ParamValue instead.
+type ArrayOrString = ParamValue
+
+// UnmarshalJSON implements the json.Unmarshaller interface.
+func (paramValues *ParamValue) UnmarshalJSON(value []byte) error {
+	// ParamValues is used for Results Value as well, the results can be any kind of
+	// data so we need to check if it is empty.
+	if len(value) == 0 {
+		paramValues.Type = ParamTypeString
+		return nil
+	}
+	if value[0] == '[' {
+		// We're trying to Unmarshal to []string, but for cases like []int or other types
+		// of nested array which we don't support yet, we should continue and Unmarshal
+		// it to String. If the Type being set doesn't match what it actually should be,
+		// it will be captured by validation in reconciler.
+		// if failed to unmarshal to array, we will convert the value to string and marshal it to string
+		var a []string
+		if err := json.Unmarshal(value, &a); err == nil {
+			paramValues.Type = ParamTypeArray
+			paramValues.ArrayVal = a
+			return nil
+		}
+	}
+	if value[0] == '{' {
+		// if failed to unmarshal to map, we will convert the value to string and marshal it to string
+		var m map[string]string
+		if err := json.Unmarshal(value, &m); err == nil {
+			paramValues.Type = ParamTypeObject
+			paramValues.ObjectVal = m
+			return nil
+		}
+	}
+
+	// By default we unmarshal to string
+	paramValues.Type = ParamTypeString
+	if err := json.Unmarshal(value, &paramValues.StringVal); err == nil {
+		return nil
+	}
+	paramValues.StringVal = string(value)
+
+	return nil
+}
+
+// MarshalJSON implements the json.Marshaller interface.
+func (paramValues ParamValue) MarshalJSON() ([]byte, error) {
+	switch paramValues.Type {
+	case ParamTypeString:
+		return json.Marshal(paramValues.StringVal)
+	case ParamTypeArray:
+		return json.Marshal(paramValues.ArrayVal)
+	case ParamTypeObject:
+		return json.Marshal(paramValues.ObjectVal)
+	default:
+		return []byte{}, fmt.Errorf("impossible ParamValues.Type: %q", paramValues.Type)
+	}
+}
+
+// NewStructuredValues creates an ParamValues of type ParamTypeString or ParamTypeArray, based on
+// how many inputs are given (>1 input will create an array, not string).
+func NewStructuredValues(value string, values ...string) *ParamValue {
+	if len(values) > 0 {
+		return &ParamValue{
+			Type:     ParamTypeArray,
+			ArrayVal: append([]string{value}, values...),
+		}
+	}
+	return &ParamValue{
+		Type:      ParamTypeString,
+		StringVal: value,
+	}
+}
+
+// NewArrayOrString is the deprecated, this is to keep backward compatibility
+var NewArrayOrString = NewStructuredValues
```

Comment:
This is not needed

Reply:
ah yes. good point. Copy and pasting issues 🤯